doc/manual/src/command-ref/cli-guideline.md

```diff
@@ -0,0 +1,571 @@
+completing an action (the group first, then the command).
+
+## Naming rules
+
+Rules are there to guide you by limiting your options. But not everything can
+fit the rules all the time. In those cases document the exceptions in [Appendix
+1: Commands naming exceptions](#appendix-1-commands-naming-exceptions) and
+provide reason. The rules want to force a Nix developer to look, not just at
+the command at hand, but also the command in a full context alongside other
+`nix` commands.
+
+```shell
+$ nix [<GROUP>] <COMMAND> [<ARGUMENTS>] [<OPTIONS>]
+```
+
+- `GROUP`, `COMMAND`, `ARGUMENTS` and `OPTIONS` should be lowercase and in a
+  singular form.
+- `GROUP` should be a  **NOUN**.
+- `COMMAND` should be a **VERB**.
+- `ARGUMENTS` and `OPTIONS` are discussed in [*Input* section](#input).
+
+## Classification
+
+Some commands are more important, some less. While we want all of our commands
+to be perfect we can only spend limited amount of time testing and improving
+them.
+
+This classification tries to separate commands in 3 categories in terms of
+their importance in regards to the new users. Users who are likely to be
+impacted the most by bad user experience. This does not mean that we will not
+take care of Tier3 commands, it means we will only put more helpful details for
+beginners into Tier1 commands.
+
+- **Tier1**: Commands used for our main use cases and most likely used by new
+  users. From Tier1 commands we expect attention to details, such as:
+    - Proper use of [colors](#colors), [emojis](#special-unicode-characters)
+      and [aligning of text](#text-alignment).
+    - [Autocomplete](#shell-completion) of options.
+    - Show [next possible steps](#next-steps).
+    - Showing some [“tips”](#educate-the-user) when running logs running tasks
+      (eg. building / downloading) in order to teach users interesting bits of
+      Nix ecosystem.
+    - [Help pages](#help-is-essential) to be as good as we can write them
+      pointing to external documentation and tutorials for more.
+- **Tier2**: Commands that are somewhere between Tier1 and Tier2, not really
+  exposing some implementation detail, but not something that we expect a user.
+  From Tier2 command we expect less attention to details, but still some:
+    - Proper use of [colors](#colors), [emojis](#special-unicode-characters)
+      and [aligning of text](#text-alignment).
+    - [Autocomplete](#shell-completion) of options.
+- **Tier3**: Commands that expose certain internal functionality of `nix`,
+  mostly used by other scripts.
+    - [Autocomplete](#shell-completion) of options.
+
+# Help is essential
+
+Help should be built into your command line so that new users can gradually
+discover new features when they need them. 
+
+## Looking for help
+
+Since there is no standard way how user will look for help we rely on ways help
+is provided by commonly used tools. As a guide for this we took `git` and
```

Comment:
Isn’t `git` notoriously famous for having a badly designed command line?

Reply:
I think the overall style of nesting and `--help` vs `-h` was popular enough, it's just the specific commands, flags, and groups that are atrocious.

Reply:
Or maybe find another reference point… By the way, Monotone supported abbreviations from command much more widely than Git (including, i.e., any unambiguous prefix which is quite convenient)

On the other hand any better reference point would lead to providing plain text output (entry per line) wheneve possible which is apparently too shell-friendly to be allowed in new Nix interface which prefers only JSON to be usable.

Reply:
Let's not forget mercurial. `hg` is known to be more user-friendly than `git` to use.